sources/lib/protobuf-tool/protobuf-tool.dylan

```diff
@@ -0,0 +1,181 @@
+  if (~generator)
+    values(#f, #())    // No generator indicates nothing changed since prev-run.
+  else
+    let (output-files, lid-file) = generate-dylan-code(generator);
+    if (lid-file)
+      maybe-update-subproject(project-file, output-files, lid-file, prev-run)
+    else
+      modify-current-project(project-file, output-files)
+    end
+  end
+end method invoke-protobuf-tool;
+
+define function maybe-update-subproject
+    (project-file :: <file-locator>, output-files :: <sequence>, lid-file :: <file-locator>,
+     prev-run :: false-or(<date>))
+ => (any-change? :: <boolean>, modified-projects :: <sequence>)
+  // Code was generated in its own library, so the only file that might be returned
+  // is the LID file.
+  if (~prev-run
+        | file-modified?(lid-file, prev-run)
+        | any?(rcurry(file-modified?, prev-run), output-files))
+    add-subproject(project-file, lid-file);
+    values(#t, vector(project-file))
+  else
+    values(#f, #())
+  end
+end function;
+
+// Code was generated without a library definition or LID file so add the output files to
+// the current project. The user might want to do this to add a method to a sealed
+// generic function, for example.
+//
+// HACK: Insert the module files after the library file, which we assume to be the first
+// file. When doing this, the user must define their library in a separate file from
+// their modules so that the module files follow the generated protobuf module files and
+// can therefore use the generated modules. What's a better way?
+// Example: the project ends up looking like this:
+//   Files: library.dylan
+//          generated-pb-module1.dylan
+//          generated-pb-module2.dylan
+//          generated-pb.dylan
+//          user-module.dylan
+//          user-main.dylan
+define function modify-current-project
+    (project-file :: <file-locator>, output-files :: <sequence>)
+ => (anything-modified? :: <boolean>, modified-projects :: <sequence> /* of: <locator> */)
+  let project = read-project-file(project-file);
+  let any-changed? = #f;
+  for (output-file in reverse(output-files))
+    let relative = relative-locator(output-file, project-file);
+    let project-files = project.project-information-files;
+    if (member?(output-file, project-files, test: \=))
+      debug-out(#"protobuf-tool", "  File already in project: %=\n", output-file);
+    else
+      project.project-information-files
+        := concatenate(list(project-files[0], output-file),
+                       copy-sequence(project-files, start: 1));
+      any-changed? := #t;
+      debug-out(#"protobuf-tool", "  Adding file to project: %=\n", output-file);
+      debug-out(#"protobuf-tool", "  Files: %=\n", project.project-information-files);
+    end;
+  end;
+  if (any-changed?)
+    write-project-file(project-file, project);
+    values(#t, vector(project-file))
+  else
+    values(#f, #())
+  end
+end function modify-current-project;
+
+/* Spec file format:
```

Comment:
Ideally this should go in the documentation rather than be left as a comment in a particular tool.

Reply:
I put the documentation in the protocol-buffers repo since we don't have a centralized place for tools in Open Dylan: https://github.com/cgay/protocol-buffers/commit/0b078ed103d00b0b4675299b8e3b63f4b30aefaa